uc.c

```diff
@@ -935,206 +918,72 @@ MemoryRegion *memory_mapping(struct uc_struct* uc, uint64_t address)
-                 }
-                 break;
-    }
-
-    return UC_ERR_OK;
-}
-
 UNICORN_EXPORT
 uc_err uc_hook_add(uc_engine *uc, uc_hook *hh, int type, void *callback, void *user_data, ...)
 {
     va_list valist;
     int ret = UC_ERR_OK;
-    int id;
-    uint64_t begin, end;
 
     va_start(valist, user_data);
 
-    if (type & UC_HOOK_MEM_READ_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_READ_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_WRITE_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_FETCH_UNMAPPED)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_FETCH_UNMAPPED, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_READ_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_READ_PROT, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_WRITE_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_PROT, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_FETCH_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_FETCH_PROT, callback, user_data, hh);
+    struct hook *hook = calloc(1, sizeof(struct hook));
+    if (hook == NULL) {
+        return UC_ERR_NOMEM;
+    }
+    hook->type = type;
+    hook->callback = callback;
+    hook->user_data = user_data;
+    *hh = (uc_hook)hook;
 
-    switch(type) {
-        default:
-            break;
-        case UC_HOOK_INTR:
-            ret = _hook_intr(uc, callback, user_data, hh);
-            break;
-        case UC_HOOK_INSN:
-            id = va_arg(valist, int);
-            ret = _hook_insn(uc, id, callback, user_data, hh);
-            break;
-        case UC_HOOK_CODE:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_code(uc, UC_HOOK_CODE, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_BLOCK:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_code(uc, UC_HOOK_BLOCK, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_MEM_READ:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_mem_access(uc, UC_HOOK_MEM_READ, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_MEM_WRITE:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_mem_access(uc, UC_HOOK_MEM_WRITE, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_MEM_READ | UC_HOOK_MEM_WRITE:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_mem_access(uc, UC_HOOK_MEM_READ | UC_HOOK_MEM_WRITE, begin, end, callback, user_data, hh);
-            break;
+    // everybody but HOOK_INSN gets begin/end, so exit early here.
```

Comment:
why dont we also give begin/end options to HOOK_INSN? i dont see why we should limit this option in this case. just give the users a choice, and they can decide when to use it?